src/java.base/share/classes/sun/security/ssl/SSLSocketImpl.java

```diff
@@ -1777,6 +1777,17 @@ private void closeSocket(boolean selfInitiated) throws IOException {
         }
 
         if (autoClose || !isLayered()) {
+            // Try to clear the kernel buffer to avoid TCP connection resets.
+            if (conContext.inputRecord instanceof SSLSocketInputRecord &&
+                isConnected) {
+                appInput.readLock.lock();
+                try {
+                    ((SSLSocketInputRecord) (conContext.inputRecord)).deplete(false);
+                } finally {
+                    appInput.readLock.unlock();
+                }
+            }
```

Comment:
The blocking on close() may be not good in practice. I would use tryLock() rather than lock() so as to avoid closure blocking.  tryLock() is not perfect, but it may be better than blocking the close().

BTW, you could use the intanceof pattern matching so as to avoid the cast (See https://openjdk.java.net/jeps/394).

```
            if (conContext.inputRecord instanceof
                                SSLSocketInputRecord inputRecord && isConnected) {
                 if (appInput.readLock.tryLock()) {
                    try {
                        inputRecord.deplete(false);
                    } finally {
                       appInput.readLock.unlock();
                    }
                 }
              }
```

Reply:
Hi Xuelei,
Thank you a lot again. Replaced lock with tryLock as you suggested